### PR TITLE
fixup! ASoC: SOF: Intel: CNL/ICL/APL: set core_get/core_put ops

### PR DESCRIPTION
**Pull request by ranj063**
Remove core_put op for CNL/ICL/APL. Multi-core with dynamic pipelines
is not supported for these platforms. So, ideally we should avoid using
this configuration in topology. Today, this results in an IPC timeout
resulting in a card registration failure during boot. Prevent this by
only keeping the core_get op to enable the core when a widget that needs
it is set up for the first time. It will be not be powered off until the
DSP enters D3.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

### CODE REVIEW

File: sound/soc/sof/intel/icl.c
Comment (plbossart) [inline sound/soc/sof/intel/icl.c:166]: for my education, is there something that would detect a bad topology with multi-cores on those platforms?

Comment (ranj063) [reply sound/soc/sof/intel/icl.c:166]: you mean in terms of core configuration as in core ID's being out of range?

Comment (plbossart) [reply sound/soc/sof/intel/icl.c:166]: possibly. We've advertised 2 or 4 cores, which is true but not supported. So somehow we need to have a 'supported_core_mask' or something that detects based on the absence of .core_get/put that only core0 can be used.

We shouldn't underestimate the creativity of our peers. Someone will try multicore on those platforms.

Comment (ranj063) [reply sound/soc/sof/intel/icl.c:166]: multi-core by itself is OK as long as they dont use it with dynamic pipelines but in any case this PR will make the option feasible. The detail that hidden is that the cores are not powered off when not needed.

Comment (plbossart) [reply sound/soc/sof/intel/icl.c:166]: ok, make sense.

**Review by ujfalusi — APPROVED** [review]
In other words, only TGL is supported.

**Comment by ranj063** [comment]
> 

TGL+ to be precise.

**Comment by keyonjie** [comment]
@ranj063 Not using get/put in pairs sounds odd to me. The ref count will keeps increasing in this case? do you mean we still support running tasks on secondary cores (specified in topology file) on CNL/ICL, and dynamic pipeline logic take place there also, the only change is that secondary cores will never be powered off except entering D3? How about D0I3 entry/exit, secondary cores need to be powered off in D0I3 no?

**Review by lyakh — APPROVED** [review]
Not sure if we can add "Tested-by" tags to GH commits, perhaps only manually, but I did test it and it fixes my problem with a modified topology on CNL.

**Comment by ranj063** [comment]
> @ranj063 Not using get/put in pairs sounds odd to me. The ref count will keeps increasing in this case? do you mean we still support running tasks on secondary cores (specified in topology file) on CNL/ICL, and dynamic pipeline logic take place there also, the only change is that secondary cores will never be powered off except entering D3? How about D0I3 entry/exit, secondary cores need to be powered off in D0I3 no?

@keyonjie your concerns are valid and this PR is intentional keeping this in mind. Yes, we can till have dynamic pipelines with components running on a secondary core. That said, whoever uses this config uses it at their own risk. It is not a configuration we support at all. As far as D0i3 is concerned, the host does nothing wrt turning off secondary cores. Should it or should it not is a separate discussion.

**Comment by keyonjie** [comment]
> > @ranj063 Not using get/put in pairs sounds odd to me. The ref count will keeps increasing in this case? do you mean we still support running tasks on secondary cores (specified in topology file) on CNL/ICL, and dynamic pipeline logic take place there also, the only change is that secondary cores will never be powered off except entering D3? How about D0I3 entry/exit, secondary cores need to be powered off in D0I3 no?
> 
> @keyonjie your concerns are valid and this PR is intentional keeping this in mind. Yes, we can till have dynamic pipelines with components running on a secondary core. That said, whoever uses this config uses it at their own risk. It is not a configuration we support at all. As far as D0i3 is concerned, the host does nothing wrt turning off secondary cores. Should it or should it not is a separate discussion.

So I would suggest not to disable multi-core on CNL/ICL/APL support like this then. BTW, I don't see any platform specific multi-core issues, am I missing something?

**Comment by ranj063** [comment]
> So I would suggest not to disable multi-core on CNL/ICL/APL support like this then. BTW, I don't see any platform specific multi-core issues, am I missing something?

@keyonjie I think you're missing th point that this PR does not disable multi-core on these platforms. on the contrary, it enables it with dynamic pipelines.

**Comment by keyonjie** [comment]
> > So I would suggest not to disable multi-core on CNL/ICL/APL support like this then. BTW, I don't see any platform specific multi-core issues, am I missing something?
> 
> @keyonjie I think you're missing th point that this PR does not disable multi-core on these platforms. on the contrary, it enables it with dynamic pipelines.

Okay, so can we keep the core_put ops for CNL/ICL/APL, but dont' really power down secondary cores on those platforms. In this case, we can keep tracking the ref_cnt correctly, and core_get/put() are used in pairs also.

File: sound/soc/sof/intel/icl.c
Comment (RanderWang) [inline sound/soc/sof/intel/icl.c:164]: remove put ?

Comment (RanderWang) [reply sound/soc/sof/intel/icl.c:164]: this comment is a copy of function name, I don't like it

**Comment by kv2019i** [comment]
@keyonjie wrote:
> Okay, so can we keep the core_put ops for CNL/ICL/APL, but dont' really power down secondary cores on those platforms. In this case, we can keep tracking the ref_cnt correctly, and core_get/put() are used in pairs also.

Your comments made me reconsider this as well. Defining a no-op hda_dsp_core_put() would be another option. This does have some side-effects:
 - sdev->enabled_cores_mask becomes inccorect and no longer matches real state of hw
 - sdev->dsp_core_ref_count is accurate with regards to how many users are using the core, but it doesn't reflect state of the actual core

I wonder if @ranj063 a good compromise would be to add a comment to snd_sof_dsp_core_put() explainin why "core_put" is optional. Essentially SOF core supports DSPs that can only:
  - power up individual core X
  - shut down all cores (at remove/suspend)

On Intel, CNL/ICL qualify to this category and not providing "core_put" is the mechanism to signal core hw is like this.

**Comment by ranj063** [comment]
@keyonjie @kv2019i I'mnot in favour of adding a dummy put ops for these platforms just to keep the core ref counts sane. This will also mess with the enabled_cores_mask and that is more problematic. The core ref_counts are strictly used only during topology elements set up/free but the enabled_cores_mask is also used beyond that during probe, dsp_init and remove.

**Comment by keyonjie** [comment]
> @keyonjie @kv2019i I'mnot in favour of adding a dummy put ops for these platforms just to keep the core ref counts sane. This will also mess with the enabled_cores_mask and that is more problematic. The core ref_counts are strictly used only during topology elements set up/free but the enabled_cores_mask is also used beyond that during probe, dsp_init and remove.

@ranj063 the only concern I have is that we might need to take the secondary core power off on CNL/ICL in some future day, so I am suggesting to go ahead with whatever verified work and convenient for doing that.

**Comment by ranj063** [comment]
> > @keyonjie @kv2019i I'mnot in favour of adding a dummy put ops for these platforms just to keep the core ref counts sane. This will also mess with the enabled_cores_mask and that is more problematic. The core ref_counts are strictly used only during topology elements set up/free but the enabled_cores_mask is also used beyond that during probe, dsp_init and remove.
> 
> @ranj063 the only concern I have is that we might need to take the secondary core power off on CNL/ICL in some future day, so I am suggesting to go ahead with whatever verified work and convenient for doing that.

@keyonjie whats been verified on CNL is turning on the cores once and leaving them on until D3 and this is what the PR achieves. If and when we enable multi-core on CNL, we can always add the put op back with the explanation that the FW supports it then

**Comment by keyonjie** [comment]
> > > @keyonjie @kv2019i I'mnot in favour of adding a dummy put ops for these platforms just to keep the core ref counts sane. This will also mess with the enabled_cores_mask and that is more problematic. The core ref_counts are strictly used only during topology elements set up/free but the enabled_cores_mask is also used beyond that during probe, dsp_init and remove.
> > 
> > 
> > @ranj063 the only concern I have is that we might need to take the secondary core power off on CNL/ICL in some future day, so I am suggesting to go ahead with whatever verified work and convenient for doing that.
> 
> @keyonjie whats been verified on CNL is turning on the cores once and leaving them on until D3 and this is what the PR achieves. If and when we enable multi-core on CNL, we can always add the put op back with the explanation that the FW supports it then

Got it. How about TGL/ADL, do we have similar issue there? We can't disable core_put there.

**Comment by ranj063** [comment]
> Got it. How about TGL/ADL, do we have similar issue there? We can't disable core_put there.

oh no no. TGL/ADL are complete in terms of multi-core implementation.

**Comment by ranj063** [comment]
@keyonjie can you please approve ths PR if you're OK with it

**Comment by keyonjie** [comment]
> @keyonjie can you please approve ths PR if you're OK with it

To be honest, I don't like the solution here, but I don't intend to block/object to it in case the change is important for multi-core + dynamic pipeline fixes. Please go without my approval.

**Pull request merged**
